redshift/redshift.go

```diff
@@ -71,149 +67,296 @@ func (c sortableColumns) Less(i, j int) bool { return c[i].Ordinal < c[j].Ordina
+	alterStmt, err := tx.Prepare(alterSQL)
+	if err != nil {
+		return fmt.Errorf("issue preparing statement: %s", err)
 	}
-	args := []interface{}{schema, table, filename, gzipSQL, jsonPaths, r.s3Info.Region}
-	log.Printf("Running command: %s with args: %v", copySQL, args)
-	_, err = tx.Stmt(copyStmt).Exec(append(args, credArgs...)...)
+
+	log.Printf("Running command: %s", alterSQL)
+	_, err = alterStmt.Exec()
 	return err
 }
 
-// RunCSVCopy copies gzipped CSV data from an S3 file into a redshift table
+// RunJSONCopy copies JSON data present in an S3 file into a redshift table.
 // this is meant to be run in a transaction, so the first arg must be a sql.Tx
-func (r *Redshift) RunCSVCopy(tx *sql.Tx, schema, table, file string, ts Table, delimiter rune, creds, gzip bool) error {
+// if not using jsonPaths, set s3File.JSONPaths to "auto"
+func (r *Redshift) RunJSONCopy(tx *sql.Tx, f s3filepath.S3File, creds, gzip bool) error {
 	var credSQL string
 	var credArgs []interface{}
 	if creds {
-		credSQL = `CREDENTIALS 'aws_access_key_id=?;aws_secret_access_key=?'`
-		credArgs = []interface{}{r.s3Info.AccessID, r.s3Info.SecretKey}
+		credSQL = `CREDENTIALS 'aws_access_key_id=%s;aws_secret_access_key=%s'`
+		credArgs = []interface{}{f.Bucket.AccessID(), f.Bucket.SecretKey()}
 	}
 	gzipSQL := ""
 	if gzip {
 		gzipSQL = "GZIP"
 	}
-
-	cols := sortableColumns{}
-	cols = append(cols, ts.Columns...)
-	sort.Sort(cols)
-	colStrings := []string{}
-	for _, ci := range cols {
-		colStrings = append(colStrings, ci.Name)
-	}
-
-	copySQL := fmt.Sprintf(`COPY "?"."?" (?) FROM '?' WITH REGION '?' ? CSV DELIMITER '?'`)
-	opts := "IGNOREHEADER 0 ACCEPTINVCHARS TRUNCATECOLUMNS TRIMBLANKS BLANKSASNULL EMPTYASNULL DATEFORMAT 'auto' ACCEPTANYDATE STATUPDATE ON COMPUPDATE ON"
-	fullCopySQL := fmt.Sprintf("%s %s %s", copySQL, opts, credSQL)
-	copyStmt, err := r.Prepare(fullCopySQL)
-	if err != nil {
-		return err
-	}
-	args := []interface{}{schema, table, strings.Join(colStrings, ", "), file, r.s3Info.Region, gzipSQL, delimiter}
-	log.Printf("Running command: %s with args: %v", fullCopySQL, args)
-	_, err = tx.Stmt(copyStmt).Exec(append(args, credArgs...)...)
+	copySQL := fmt.Sprintf(`COPY "%s"."%s" FROM '%s' WITH %s JSON '%s' REGION '%s' TIMEFORMAT 'auto' STATUPDATE ON COMPUPDATE ON %s`, f.Schema, f.Table, f.GetDataFilename(), gzipSQL, f.JSONPaths, f.Bucket.Region(), credSQL)
+	fullCopySQL := fmt.Sprintf(fmt.Sprintf(copySQL, credArgs...))
+	log.Printf("Running command: %s", copySQL)
+	// can't use prepare b/c of redshift-specific syntax that postgres does not like
+	_, err := tx.Exec(fullCopySQL)
 	return err
 }
 
 // RunTruncate deletes all items from a table, given a transaction, a schema string and a table name
 // you shuold run vacuum and analyze soon after doing this for performance reasons
 func (r *Redshift) RunTruncate(tx *sql.Tx, schema, table string) error {
-	truncStmt, err := r.Prepare(`DELETE FROM "?"."?"`)
+	truncStmt, err := tx.Prepare(`DELETE FROM "?"."?"`)
```

Comment:
This one shouldn't need to be a const, but worth considering for consistent styling